cluster-autoscaler/cloudprovider/openstack/openstack_manager_heat.go

```diff
@@ -0,0 +1,299 @@
+	}
+
+	err = osm.UpdateNodeCount(nodegroup, updatedNodeCount)
+	if err != nil {
+		return fmt.Errorf("could not set new cluster size: %v", err)
+	}
+	return nil
+}
+
+// GetClusterStatus returns the current status of the magnum cluster.
+func (osm *OpenstackManagerHeat) GetClusterStatus() (string, error) {
+	cluster, err := clusters.Get(osm.clusterClient, osm.clusterName).Extract()
+	if err != nil {
+		return "", fmt.Errorf("could not get cluster: %v", err)
+	}
+	return cluster.Status, nil
+}
+
+// CanUpdate checks if the cluster status is present in a set of statuses that
+// prevent the cluster from being updated.
+// Returns if updating is possible and the status for convenience.
+func (osm *OpenstackManagerHeat) CanUpdate() (bool, string, error) {
+	clusterStatus, err := osm.GetClusterStatus()
+	if err != nil {
+		return false, "", fmt.Errorf("could not get cluster status: %v", err)
+	}
+	return !StatusesPreventingUpdate.Has(clusterStatus), clusterStatus, nil
+}
+
+// GetStackStatus returns the current status of the heat stack used by the magnum cluster.
+func (osm *OpenstackManagerHeat) GetStackStatus() (string, error) {
+	stack, err := stacks.Get(osm.heatClient, osm.stackName, osm.stackID).Extract()
+	if err != nil {
+		return "", fmt.Errorf("could not get stack from heat: %v", err)
+	}
+	return stack.Status, nil
+}
+
+// WaitForStackStatus checks once per second to see if the heat stack has entered a given status.
+// waits for heat stack to change to a given status.
+// Returns when the status is observed or the timeout is reached.
+func (osm *OpenstackManagerHeat) WaitForStackStatus(status string, timeout time.Duration) error {
+	klog.V(2).Infof("Waiting for stack %s status", status)
+	for start := time.Now(); time.Since(start) < timeout; time.Sleep(osm.waitTimeStep) {
+		currentStatus, err := osm.GetStackStatus()
+		if err != nil {
+			return fmt.Errorf("error waiting for stack status: %v", err)
+		}
+		if currentStatus == status {
+			klog.V(0).Infof("Waited for stack %s status", status)
+			return nil
+		}
+	}
+	return fmt.Errorf("timeout (%v) waiting for stack status %s", timeout, status)
+}
+
+// GetStackID gets the heat stack ID from the magnum cluster.
+func (osm *OpenstackManagerHeat) GetStackID() (string, error) {
+	cluster, err := clusters.Get(osm.clusterClient, osm.clusterName).Extract()
+	if err != nil {
+		return "", fmt.Errorf("could not get cluster: %v", err)
+	}
+	klog.V(0).Infof("Cluster stack ID: %s", cluster.StackID)
+	return cluster.StackID, nil
+}
+
+// GetStackName lists all heat stacks and finds the one with a matching ID
+// to the heat stack ID obtained from the magnum cluster.
+func (osm *OpenstackManagerHeat) GetStackName() (string, error) {
+	// Calls to stacks.Get require both ID and name, so to get the name
+	// we have to inspect all stacks and find the one with the matching ID.
+
+	allStacksPages, err := stacks.List(osm.heatClient, nil).AllPages()
```

Comment:
could this just be something like
```
stacks.Find(osm.heatClient, osm.stackID).Extract()
```

Reply:
This was added to gophercloud after the commit that I have had to use. When kubernetes drops the old version of [cloudprovider/openstack](https://github.com/kubernetes/kubernetes/tree/master/pkg/cloudprovider/providers/openstack) we can upgrade to a more recent version.

Reply:
that makes sense, thanks for the info

Reply:
Now that we have this version, this is done in https://github.com/kubernetes/autoscaler/pull/1690/commits/0d7387f3806a8c9f48f60030214a57b2bf1282b6.